test/unit/communication/test_all_reduce.cpp

```diff
@@ -0,0 +1,200 @@
+//
+// Distributed Linear Algebra with Future (DLAF)
+//
+// Copyright (c) 2018-2019, ETH Zurich
+// All rights reserved.
+//
+// Please, refer to the LICENSE file in the root directory.
+// SPDX-License-Identifier: BSD-3-Clause
+//
+
+#include "dlaf/communication/sync/all_reduce.h"
+
+#include <gtest/gtest.h>
+
+#include "dlaf/common/data_descriptor.h"
+#include "dlaf/communication/communicator_grid.h"
+
+#include "dlaf_test/helper_communicators.h"
+#include "dlaf_test/util_types.h"
+
+using namespace dlaf;
+using namespace dlaf_test;
+using namespace dlaf::comm;
+
+using AllReduceTest = SplittedCommunicatorsTest;
+
+template <class T, class = std::enable_if_t<std::is_arithmetic<T>::value>>
+void DLAF_EXPECT_NEAR(const T a, const T b, const T abs_err) {
+  EXPECT_NEAR(a, b, abs_err);
```

Comment:
Using the macro in a separate function completely mess up the output message in case of error. How do you deal with this problem?

Reply:
I didn't think about this.

I'm going to check if `AssertionResult` can be useful in this context (or if something else exists for this purpose).

Reply:
What about using always `EXPECT_LE(std::abs(a-b), error)`?

Reply:
FIY, in this specific instance, we can easily use your proposal (it does not need to be generic, because this test runs just for `TypeParam = std::complex<double>`. IIRC, this is because it is not possible to mix TestFixtures and TypedTests in GTEST (AFAIK, I have to check if they improved it).

However, I did a quick test with `AssertionResult` and it works nicely. It is possible to define a function and then check its result with `EXPECT_TRUE`.

```
template <class T, class U>
::testing::AssertionResult DLAF_EXPECT_NEAR(const T& a, const T& b, const U& abs_err) {
  if (std::abs(a-b) < abs_err)
    return ::testing::AssertionSuccess();
  return ::testing::AssertionFailure() << "They are not equal " << a << " and " << b;
}
```

The message construction is optional, but then testing it with `EXPECT_TRUE(DLAF_EXPECT_NEAR(a, b, err))` allows to report the error in the right source location.

```
dla-future/test/unit/communication/test_all_reduce.cpp:209: Non-fatal failure:
Value of: DLAF_EXPECT_NEAR(value * static_cast<TypeParam>(NUM_MPI_RANKS), data_strided[mem_pos], NUM_MPI_RANKS * TypeUtilities<TypeParam>::error)
  Actual: false (They are not equal (52,104) and (52,104))
Expected: true
```

It may be an overshoot for this particular instance of the problem, but I thought it was worth sharing it.

Reply:
Note that `EXPECT_LE(std::abs(a-b), error)` works both for real and complex cases:
https://en.cppreference.com/w/cpp/numeric/math/fabs
https://en.cppreference.com/w/cpp/numeric/complex/abs